File: activedirectory/list_users.rb
@log.trace("Started execution 'activedirectory:list_users.rb' flintbit...") # execution Started
begin
    # Flintbit input parametes
    # Mandatory
    @connector_name = @input.get('connector_name') # active-directory connector name
    @dn = @input.get('distinguished-names') #distinguished-names to list the user in the active-directory

    # Optional
    request_timeout = @input.get('timeout')	# Execution time of the Flintbit in milliseconds (default timeout is 60000 milloseconds)

    connector_call = @call.connector('testad')
                          .set('action', 'list-users')
                          .set('distinguished-names', @dn)


    if @connector_name.nil? || @connector_name.empty?
        raise 'Please provide "LDAP connector name (connector_name)" to list users'
    end

    if @dn.nil? || @dn.empty?
        raise 'Please provide "distinguished name (distinguished-names)" to list users'
    end


    if request_timeout.nil? || request_timeout.is_a?(String)
        @log.trace("Calling #{@connector_name} with default timeout...")
        # calling active-directory connector
        response = connector_call.sync
    else
        @log.trace("Calling #{@connector_name} with given timeout #{request_timeout}...")
        # calling active-directory connector
        response = connector_call.timeout(request_timeout).sync
    end

    # active-directory  Connector Response Meta Parameters
    response_exitcode = response.exitcode # Exit status code
    response_message =  response.message # Execution status message

    @log.info("RESPONSE :: #{response}")

    if response_exitcode == 0
        @log.info("Success in executing #{@connector_name} Connector, where exitcode :: #{response_exitcode} | message :: #{response_message}")
        @output.set('result', response.to_s).set('exit-code', 0)
    else
        @log.error("ERROR in executing #{@connector_name} where, exitcode :: #{response_exitcode} | message :: #{response_message}")
        @output.set('message', response_message.to_s).set('exit-code', -1)
    end

  rescue Exception => e
    @log.error(e.message)
    @output.set('exit-code', -1).set('message', e.message)
end

@log.trace("Finished execution 'activedirectory:list_users.rb' flintbit...") # Execution Finished

File: azure_ad/operations/list_users.js
log.trace("Started executing 'flint-util:azure_ad:operations:list_users.js' flintbit")

log.trace("Inputs for 'flint-util:azure_ad:operations:list_users.js' :: " + input)
action = "list-users"


// Input clone
input_clone = JSON.parse(input)

// Initializaing vars
client_id = ""
tenant_id = ""
key = ""
subscription_id = ""
ms_azure_parameters = ""
connector_name = ""

// Check if service params for Azure exist
if (input_clone.hasOwnProperty('ms_azure_parameters')) {
    ms_azure_parameters = input.get('ms_azure_parameters')

    // Client ID
    if (!input_clone.hasOwnProperty('client_id')) {
        client_id = ms_azure_parameters.get('client_id')
        log.trace("Client ID taken from service parameters")
    }

    // Tenant ID
    if (!input_clone.hasOwnProperty('tenant_id')) {
        tenant_id = ms_azure_parameters.get('tenant_id')
        log.trace("Tenant ID taken from service parameters")
    }

    // Key
    if (!input_clone.hasOwnProperty('key')) {
        key = ms_azure_parameters.get('key')
        log.trace("Key taken from service parameters")
    }

    // Subscription ID
    if (!input_clone.hasOwnProperty('subscription_id')) {
        subscription_id = ms_azure_parameters.get('subscription_id')
        log.trace("Subscription ID taken from service parameters")
    }
    // Connector Name
    if (!input_clone.hasOwnProperty('connector_name')) {
        connector_name = ms_azure_parameters.get('connector_name')
        log.trace("Connector name is given in service parameter: "+connector_name)
    }
} else {
    log.info("Optional azure service parameters are not present")

    client_id = input.get('client_id')
    log.trace("Client ID: "+client_id)
    tenant_id = input.get('tenant_id')
    log.trace("Tenant ID is given")
    key = input.get('key')
    log.trace("Key is given")
    subscription_id = input.get('subscription_id')
    log.trace("Subscription ID: "+subscription_id)
    connector_name = input.get('connector_name')
    log.trace("Connector name is "+connector_name)
}

// Connector call
log.trace("Calling MS Azure connector for action: " + action)

// Active Directory Domain - Mandatory
active_directory_domain = ""
if(input_clone.hasOwnProperty('active_directory_domain')){
    log.info("Active directory domain is given")
    active_directory_domain = input.get('active_directory_domain')
    if (active_directory_domain != null || active_directory_domain != "") {
        active_directory_domain = input.get('active_directory_domain')
        log.trace("Azure Active directory domain is "+ active_directory_domain)
    }else{
        log.error("Active directory domain is null or empty")
    }
}else{
    log.error("Active directory domain is not given")
}

connector_call = call.connector(connector_name)
    .set('action', action)
    .set('client-id', client_id)
    .set('tenant-id', tenant_id)
    .set('key', key)
    .set('subscription-id', subscription_id)
    .set('active-directory-domain' , active_directory_domain)
    .timeout(120000)
    .sync()

    log.trace("Connector response: "+connector_call)

    exit_code = connector_call.exitcode()
    message = connector_call.message()


    if(exit_code == 0){
        log.trace(connector_call)
        output.set("result", JSON.parse(connector_call))
    }else{
        log.trace("Error: "+message)
        output.set("error", message)
    }

    log.trace("Finished executing 'flint-util:azure_ad:operations:list_users.js' flintbit")